include/pagmo/algorithms/gwo.hpp

```diff
@@ -0,0 +1,362 @@
+        const auto bounds = prob.get_bounds();
+        const auto &lb = bounds.first;
+        const auto &ub = bounds.second;
+        auto NP = pop.size(); // number of agent(wolf) equal to population size
+        auto count = 1u;      // regulates the screen output
+
+        // PREAMBLE-------------------------------------------------------------------------------------------------
+        // We start by checking that the problem is suitable for this
+        // particular algorithm.
+        if (prob.get_nc() != 0u) {
+            pagmo_throw(std::invalid_argument, "Non linear constraints detected in " + prob.get_name() + " instance. "
+                                                   + get_name() + " cannot deal with them");
+        }
+        if (prob_f_dimension != 1u) {
+            pagmo_throw(std::invalid_argument, "Multiple objectives detected in " + prob.get_name() + " instance. "
+                                                   + get_name() + " cannot deal with them");
+        }
+        if (prob.is_stochastic()) {
+            pagmo_throw(std::invalid_argument,
+                        "The problem appears to be stochastic " + get_name() + " cannot deal with it");
+        }
+        // Get out if there is nothing to do.
+        if (m_gen == 0u) {
+            return pop;
+        }
+        if (pop.size() < 3u) {
+            pagmo_throw(std::invalid_argument, get_name() + " needs at least 3 individuals in the population, "
+                                                   + std::to_string(pop.size()) + " detected");
+        }
+        // ---------------------------------------------------------------------------------------------------------
+
+        // No throws, all valid: we clear the logs
+        m_log.clear();
+
+        // Some vectors used during evolution are declared.
+        vector_double alpha_pos(dim);
+        vector_double beta_pos(dim);
+        vector_double delta_pos(dim);
+        vector_double a_vector(3); // vector coefficient for encircling prey
+        vector_double c_vector(3); // vector coefficient for encircling prey
+        vector_double d_vector(3); // position vector for alpha, beta and delta
+        vector_double x_vector(3); // vector used to calculate position of an agent in a single dimension
+        double a;                  // coefficient which decrease linearly from 2 to 0 over generations
+        double r1, r2;             // random coefficient between 0 and 1;
+        double alpha_score = std::numeric_limits<double>::infinity();
+        double beta_score = std::numeric_limits<double>::infinity();
+        double delta_score = std::numeric_limits<double>::infinity();
+
+        // assume pop.get_x() return vector<vector_double>
+        auto agents_position = pop.get_x();
+        std::uniform_real_distribution<double> drng(0., 1.); // to generate a number in [0, 1)
+
+        // Main gwo iterations
+        for (decltype(m_gen) gen = 1u; gen <= m_gen; ++gen) {
```

Comment:
Hi.. getting the best three individuals at the initialization, in the first part main gwo iterations from line 164 to line 178 it is already done here...
if we initialize it with best three individuals before gwo iteration, these code will run for nothing or we add a condition that it shouldn't run in the first iteration which is unnecessary

Reply:
The way its currently written is incorrect. Upon entering the main loop alpha beta and gamma need to be defined and not contain zeros and infinities. Otherwise on line 198, for example, and during the first iteration (i=0), beta will contain garbage.

I agree that the update of alpha beta and gamma can be done later in the loop, after line 214.

Reply:
You are right, the performance of the algorithm seem to be increase after this changes!